app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php

```diff
@@ -174,7 +170,54 @@ public function getCartDataJs($multipage = true)
 
             // Call Bolt create order API
             try {
-                $orderCreationResponse = $this->createBoltOrder($quote, $multipage);
+                /////////////////////////////////////////////////////////////////////////////////
+                // We create a copy of the quote that is immutable by the customer/frontend
+                // Bolt saves this quote to the database at Magento-side order save time.
+                // This assures that the quote saved to Magento matches what is stored on Bolt
+                // Only shipping, tax and discounts can change, and only if the shipping, tax
+                // and discount calculations change on the Magento server
+                ////////////////////////////////////////////////////////////////////////////////
+
+                /*********************************************************/
+                /* Clean up resources that may have previously been saved
+                /* @var Mage_Sales_Model_Quote[] $expired_quotes */
+                $expired_quotes = Mage::getModel('sales/quote')
+                    ->getCollection()
+                    ->addFieldToFilter('parent_quote_id', $quote->getId());
+
+                foreach( $expired_quotes as $expired_quote) {
+                    $expired_quote->delete();
+                }
+                /*********************************************************/
+
+                /* @var Mage_Sales_Model_Quote $immutableQuote */
+                $immutableQuote = Mage::getSingleton('sales/quote');
+
+                if (!$multipage) {
+                    // For the checkout page we want to set the
+                    // billing and shipping, and shipping method at this time.
+                    // For multi-page, we add the addresses during the shipping and tax hook
+                    // and the chosen shipping method at order save time.
+                    $immutableQuote
+                        ->setBillingAddress($quote->getBillingAddress())
+                        ->setShippingAddress($quote->getShippingAddress())
+                        ->getShippingAddress()
+                            ->setShippingMethod($quote->getShippingAddress()->getShippingMethod())
+                            ->save();
+                }
+
+                $immutableQuote
+                    ->setCustomer($quote->getCustomer())
+                    ->merge($quote)
```

Comment:
I'm not familiar with the merge function. Does this take anything from the quote that is passed in and assign anything that isn't already set on the immutableQuote? Just curious.

Reply:
```
    /**
     * Merge quotes
     *
     * @param   Mage_Sales_Model_Quote $quote
     * @return  Mage_Sales_Model_Quote
     */
    public function merge(Mage_Sales_Model_Quote $quote)
    {
        Mage::dispatchEvent(
            $this->_eventPrefix . '_merge_before',
            array(
                 $this->_eventObject=>$this,
                 'source'=>$quote
            )
        );

        foreach ($quote->getAllVisibleItems() as $item) {
            $found = false;
            foreach ($this->getAllItems() as $quoteItem) {
                if ($quoteItem->compare($item)) {
                    $quoteItem->setQty($quoteItem->getQty() + $item->getQty());
                    $found = true;
                    break;
                }
            }

            if (!$found) {
                $newItem = clone $item;
                $this->addItem($newItem);
                if ($item->getHasChildren()) {
                    foreach ($item->getChildren() as $child) {
                        $newChild = clone $child;
                        $newChild->setParentItem($newItem);
                        $this->addItem($newChild);
                    }
                }
            }
        }

        /**
         * Init shipping and billing address if quote is new
         */
        if (!$this->getId()) {
            $this->getShippingAddress();
            $this->getBillingAddress();
        }

        if ($quote->getCouponCode()) {
            $this->setCouponCode($quote->getCouponCode());
        }

        Mage::dispatchEvent(
            $this->_eventPrefix . '_merge_after',
            array(
                 $this->_eventObject=>$this,
                 'source'=>$quote
            )
        );

        return $this;
    }
```

Reply:
```
    /**
     * Merge quotes
     *
     * @param   Mage_Sales_Model_Quote $quote
     * @return  Mage_Sales_Model_Quote
     */
    public function merge(Mage_Sales_Model_Quote $quote)
    {
        Mage::dispatchEvent(
            $this->_eventPrefix . '_merge_before',
            array(
                 $this->_eventObject=>$this,
                 'source'=>$quote
            )
        );

        foreach ($quote->getAllVisibleItems() as $item) {
            $found = false;
            foreach ($this->getAllItems() as $quoteItem) {
                if ($quoteItem->compare($item)) {
                    $quoteItem->setQty($quoteItem->getQty() + $item->getQty());
                    $found = true;
                    break;
                }
            }

            if (!$found) {
                $newItem = clone $item;
                $this->addItem($newItem);
                if ($item->getHasChildren()) {
                    foreach ($item->getChildren() as $child) {
                        $newChild = clone $child;
                        $newChild->setParentItem($newItem);
                        $this->addItem($newChild);
                    }
                }
            }
        }

        /**
         * Init shipping and billing address if quote is new
         */
        if (!$this->getId()) {
            $this->getShippingAddress();
            $this->getBillingAddress();
        }

        if ($quote->getCouponCode()) {
            $this->setCouponCode($quote->getCouponCode());
        }

        Mage::dispatchEvent(
            $this->_eventPrefix . '_merge_after',
            array(
                 $this->_eventObject=>$this,
                 'source'=>$quote
            )
        );

        return $this;
    }
```